### DISCUSSION TITLE
Smallstep internal CA set up but no joy from win acme

### DISCUSSION THREAD

**Discussion by CvW** [smallstep/certificates#592] [Q&A] [answered]
Good morning.
I have successfully set up the small step CA on a pi, added the root trust certificates manually to the windows certificates manager, and entering https://tinyca.internal/acme/acme/directory returns the following with the trusted symbol in MS edge:

`{"newNonce":"https://tinyca.internal/acme/acme/new-nonce","newAccount":"https://tinyca.internal/acme/acme/new-account","newOrder":"https://tinyca.internal/acme/acme/new-order","revokeCert":"https://tinyca.internal/acme/acme/revoke-cert","keyChange":"https://tinyca.internal/acme/acme/key-change"}`

So far, everything suggests that I should be able to be able to use win acme to supply my windows machine with a certificate. Since I continue to fail at this, I decided to run WACS in admin  + verbose mode to hopefully resolve this issue. Here are my steps and the results. 

`C:\Program Files\Win ACME>wacs.exe --verbose
 [VERB] Verbose mode logging enabled
 [VERB] ExePath: C:\Program Files\Win ACME\wacs.exe
 [VERB] ResourcePath: C:\Program Files\Win ACME\
 [VERB] PluginPath: C:\Program Files\Win ACME\
 [VERB] Looking for settings.json in C:\Program Files\Win ACME\
 [DBUG] Config folder: C:\ProgramData\win-acme\acme-v02.api.letsencrypt.org
 [DBUG] Log path: C:\ProgramData\win-acme\acme-v02.api.letsencrypt.org\Log
 [DBUG] Cache path: C:\ProgramData\win-acme\acme-v02.api.letsencrypt.org\Certificates
 [DBUG] secrets.json not found
 [VERB] Arguments: --verbose
 [DBUG] Renewal period: 55 days
 [VERB] Sending e-mails False

 [INFO] A simple Windows ACMEv2 client (WACS)
 [INFO] Software version 2.1.17.1065 (release, pluggable, standalone, 64-bit)
 [INFO] Connecting to https://acme-v02.api.letsencrypt.org/...
 [VERB] SecurityProtocol setting: SystemDefault
 [DBUG] Send GET request to https://acme-v02.api.letsencrypt.org/directory
 [VERB] Request completed with status OK
 [DBUG] Connection OK!
 [DBUG] IIS not detected
 [DBUG] Running with administrator credentials
 [WARN] Scheduled task not configured yet
 [INFO] Please report issues at https://github.com/win-acme/win-acme
 [VERB] Unicode display test: Chinese/語言 Russian/язык Arab/لغة

 N: Create certificate (default settings)
 M: Create certificate (full options)
 R: Run renewals (0 currently due)
 A: Manage renewals (0 total)
 O: More options...
 Q: Quit

 Please choose from the menu: M

 [INFO] Running in mode: Interactive, Advanced
 [VERB] Adding 8.8.8.8 as DNS server
 [VERB] Adding 1.1.1.1 as DNS server
 [VERB] Adding 8.8.4.4 as DNS server
 [WARN] Target plugin IIS not available: No supported version of IIS detected.

  Please specify how the list of domain names that will be included in the
  certificate should be determined. If you choose for one of the "all bindings"
  options, the list will automatically be updated for future renewals to
  reflect the bindings at that time.

 1: Read site bindings from IIS
 2: Manual input
 3: CSR created by another program
 C: Abort

 How shall we determine the domain(s) to include in the certificate?: 2

 Enter comma-separated list of host names, starting with the common name: nuc.internal
 [INFO] Target generated using plugin Manual: nuc.internal
 Suggested friendly name '[Manual] nuc.internal', press <Enter> to accept or type an alternative: <Enter>

  The ACME server will need to verify that you are the owner of the domain
  names that you are requesting the certificate for. This happens both during
  initial setup *and* for every future renewal. There are two main methods of
  doing so: answering specific http requests (http-01) or create specific dns
  records (dns-01). For wildcard domains the latter is the only option. Various
  additional plugins are available from https://github.com/win-acme/win-acme/.

 1: [http-01] Save verification files on (network) path
 2: [http-01] Serve verification files from memory
 3: [http-01] Upload verification files via FTP(S)
 4: [http-01] Upload verification files via SSH-FTP
 5: [http-01] Upload verification files via WebDav
 6: [dns-01] Create verification records manually (auto-renew not possible)
 7: [dns-01] Create verification records with acme-dns (https://github.com/joohoi/acme-dns)
 8: [dns-01] Create verification records with your own script
 9: [tls-alpn-01] Answer TLS verification request from win-acme
 C: Abort

 How would you like prove ownership for the domain(s)?: 7

 URL of the acme-dns server: https://tinyca.internal/acme/acme/directory
 [DBUG] Using C:\ProgramData\win-acme\acme-dns\tinyca.internalacmeacmedirectory for acme-dns configuration
 [INFO] Creating new acme-dns registration for domain nuc.internal
 [DBUG] Send POST request to https://tinyca.internal/acme/acme/register
 **[VERB] Request completed with status NotFound**
 [EROR] Error creating acme-dns registration
Newtonsoft.Json.JsonSerializationException: Error converting value 404 to type 'PKISharp.WACS.Clients.AcmeDnsClient+RegisterResponse'. Path '', line 1, position 3.
 ---> System.ArgumentException: Could not cast or convert from System.Int64 to PKISharp.WACS.Clients.AcmeDnsClient+RegisterResponse.
   at Newtonsoft.Json.Utilities.ConvertUtils.EnsureTypeAssignable(Object value, Type initialType, Type targetType)
   at Newtonsoft.Json.Utilities.ConvertUtils.ConvertOrCast(Object initialValue, CultureInfo culture, Type targetType)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.EnsureType(JsonReader reader, Object value, CultureInfo culture, JsonContract contract, Type targetType)
   --- End of inner exception stack trace ---
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.EnsureType(JsonReader reader, Object value, CultureInfo culture, JsonContract contract, Type targetType)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonSerializer.Deserialize(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value)
   at PKISharp.WACS.Clients.AcmeDnsClient.Register()
 [EROR] Validation plugin acme-dns was unable to generate options`

The thing that jumps out at me is the POST request that doesn't seem to be able to reach the "register" on the tinyca. Is that behavior expected? What else could I be doing wrong?

**Comment by bfqrst** [comment]
Hey @CvW,

I have been successfully able to get certs with this combo BUT I remember one thing that bit me:

If you put in your custom ACME endpoint omit the word `directory` at the end, because win-acme appends that by default. Took me a while to notice this. I'm not sure that that's your error, but give it a try!

Cheers!

**Comment by CvW** [comment, marked as answer]
Thank you so much for helping. Unfortunately, that didn't do it either. I entered the URL two different ways, once with the last directory slash, once without. The results are the same, see below:

` URL of the acme-dns server: https://tinyca.internal/acme/acme

 [DBUG] Using C:\ProgramData\win-acme\acme-dns\tinyca.internalacmeacme for acme-dns configuration
 [INFO] Creating new acme-dns registration for domain nuc.internal
 [DBUG] Send POST request to https://tinyca.internal/acme/register
 [VERB] Request completed with status NotFound
 [EROR] Error creating acme-dns registration
Newtonsoft.Json.JsonSerializationException: Error converting value 404 to type 'PKISharp.WACS.Clients.AcmeDnsClient+RegisterResponse'. Path '', line 1, position 3.
 ---> System.ArgumentException: Could not cast or convert from System.Int64 to PKISharp.WACS.Clients.AcmeDnsClient+RegisterResponse.`

Or, with the second slash (which seems appropriate)

` URL of the acme-dns server: https://tinyca.internal/acme/acme/

 [DBUG] Using C:\ProgramData\win-acme\acme-dns\tinyca.internalacmeacme for acme-dns configuration
 [INFO] Creating new acme-dns registration for domain nuc.internal
 [DBUG] Send POST request to https://tinyca.internal/acme/acme/register
 [VERB] Request completed with status NotFound
 [EROR] Error creating acme-dns registration
Newtonsoft.Json.JsonSerializationException: Error converting value 404 to type 'PKISharp.WACS.Clients.AcmeDnsClient+RegisterResponse'. Path '', line 1, position 3.
 ---> System.ArgumentException: Could not cast or convert from System.Int64 to PKISharp.WACS.Clients.AcmeDnsClient+RegisterResponse.` Etc.

It's a bit discouraging.

**Reply by bfqrst** [reply]
Don't be discouraged, the step toolchain is really good, this seems to be some kind of misconception and/or misconfiguration... Let's see...

From what I see, you chose to complete the ACME DNS challenge. Couple of questions regarding this:

1. Does `nuc.internal` resolve back to the correct IP?
2. Is your DNS ready to programmatically alter records?

I'd keep it really really basic for the time being and try the HTTP challenge. Make sure port 80 (or whatever port you choose is not occupied though). Also make sure the directory from where you start the command is writable...

Good luck!

**Reply by CvW** [reply]
Thank you, I will try the HTTP challenge next. In the meantime, here are the results for both the CA as well as the machine I'm trying to get  acme to run on:
`C:\Program Files\Win ACME>nslookup tinyca.internal
Server:  pi244.internal
Address:  192.168.1.244

Name:    tinyca.internal
Address:  192.168.1.247


C:\Program Files\Win ACME>nslookup nuc.internal
Server:  pi244.internal
Address:  192.168.1.244

Name:    nuc.internal
Address:  192.168.1.190`

**Reply by CvW** [reply]
"Is your DNS ready to programmatically alter records?" 
That  is a great  question to which I have no answer, showing  my ignorance.  I can certainly alter the DNS  records locally, by hand. Not sure how that would work programmatically. 

As for HTTP request, is there a tutorial you can point me to? None of the options seem to include using a Step CA. 

Apologies again. I'm wondering if it doesn't make more sense to take advantage of external resources (i.e. my registered domains) and forgo the use of an internal CA altogether. Too much of a dolt I am.

**Reply by tashian** [reply]
Hi @CvW! Carl from Smallstep here. We haven't seen many people running `wacs.exe` to get certificates from `step-ca` yet. We'd love to add `wacs.exe` to our [ACME client usage documentation](https://smallstep.com/docs/tutorials/acme-protocol-acme-clients). There's no reason it shouldn't work, with DNS-01 or HTTP-01 validation. ACME is ACME and `step-ca` is compliant with every ACME client we've tried so far. But sometimes configuring clients to trust an internal ACME CA can be tricky.

For example, with [`certbot`](https://certbot.eff.org/), you have to set a `REQUESTS_CA_BUNDLE` environment variable pointing to your `root_ca.crt`.

The 404 from win-acme is confusing though, because that doesn't sound like a trust / certificate validation issue.

I couldn't find anything in win-acme's manual about trust settings. @WouterTinus I wonder if you have any insights into what's going on here? Does `win-acme` use the system trust store on Windows? Or, does a private CA need to be explicitly trusted?

**Reply by WouterTinus** [reply]
Hi @tashian, happy to work with you here. It seems that @CvW has a couple of thinks mixed up here. 

You are calling into the default Let's Encrypt ACME endpoint instead of your own one. The default endpoint can be changed in the `settings.json` file that comes with the program. If you don't specify it there, you have to provide it using the argument `--baseuri`. 

Then as for validation method you picked `acme-dns`, which is a whole different tool (see https://github.com/joohoi/acme-dns - that link is also in the menu). But instead of pointing to an acme-dns instance, you point to the step-ca endpoint.

Obviously step-ca doesn't understand acme-dns commands and Let's Encrypt cannot do anything with step-ca, so it all comes crashing down 😄 

So in summary: use `--baseuri your.own.url` and pick a different validation method. Preferably the default one, or at least an http-01 flavor because they are generally easier to do than DNS validation.

**Reply by tashian** [reply]
@CvW given @WouterTinus's feedback, in addition to setting the `baseuri`, I'd suggest trying option 9 (`tls-alpn-01`) for the ACME challenge.

**Reply by CvW** [reply]
You guys are simply amazing. Thank you so much! Once I used the --baseuri https://tinyca.internal/acme/acme/ argument as part of my wacs.exe startup, everything started to be happy. My suggestion to the smallstep team would be for that  nugget of info to be included in the instructions for Windows users since WACS seems to be pretty popular on that platform. 

A huge thank you also to everyone at WACS for offering such a verbose mode. It makes all the difference in the world when a app can tell you exactly where things went off the  rails. Or in my case, it dumbfounded me but was illuminating to you! :)

You cannot imagine my happiness as I selected option 9 and saw everything execute happily where before my ignorance caused everything to crash.  Now I can launch my Windows remote session without  the application complaining that the connection cannot be trusted. 

FWIW, I did manually transfer the root CA CRT file into the windows trust store. I don't know if that helped, was necessary, etc. But if this step is needed, transferring said file is pretty trivial. 

Next, I will try  to do the same thing from FreeBSD, RouterOS (whatever that is based on!), Raspian, and so on. Many more platforms to learn... but I could not have done it without  all of you. So thank you and I will turn around and make a modest donation to both smallstep and WinACME to support this important work. 

As part of the amazing verbose feedback I also saw the auto-scheduled renewal task, which is absolutely perfect. Just a quick question -  does WinACME have to be open in the background? I ask since I wonder if task will attempt to renew with  default lets encrypt server settings vs. using the baseuri argument. You  likely thought of this a long time ago, but I'm just a little paranoid now because I've failed at all this for so long. 

Again,  thank you all so much, I really really appreciate this help.

**Reply by tashian** [reply]
Great! I'm glad you got this resolved. We can't accept donations but we'll gladly accept any promotion of `step-ca` that you can offer, even if it's just a tweet (we're [@smallsteplabs on twitter](https://twitter.com/smallsteplabs)).

Finally, I've created smallstep/docs#87 to add `win-acme` to our documentation.

**Reply by CvW** [reply]
I don't have a twitter account, sorry. I'll try to promote you however over at the TrueNAS user forums, where I am pretty  active.

**Reply by WouterTinus** [reply]
Happy to help!

**Reply by whoang44** [reply]
Awesome thread. Without the help of you guys, we would not be able to make this thing work. This tread should be posted on small step Q&A. They don't have much info for Windows clients

**Reply by tashian** [reply]
@whoang44 glad it worked for you! I've added it to our [list of community-contributed tutorials and workflows](https://github.com/smallstep/certificates/discussions/765)
